фз 10

Maks1-15 opened the pull request:
Добавлена необходимая функциональность
Запись осуществляется в базу данных
view в процессе разработки
unit тесты в процессе разработки(пока не сильно знаю, как их именно писать, тестировал через postman, вроде все классно)

Maks1-15 committed:
фз 10
Добавлена необходимая функциональность
Запись осуществляется в базу данных
view в процессе разработки
unit тесты в процессе разработки(пока не сильно знаю, как их именно писать, тестировал через postman, вроде все классно)

afomkina commented:
Сейчас не проходят проверки, если не очень понятно почему - стоит написать наставнику и попросить помощи

Maks1-15 committed:
фз 10
Доработал функционал, добавил 2 своих аннотации

Тесты не написал, не сильно разобрался с ними, но смотрел по postman'у, все сохраняется
Добавил логги
Оставил вопросы по коду в комментариях

Maks1-15 committed:
фз 10
Исправление внутренних тестов

Maks1-15 committed:
фз 10
Исправление внутренних тестов

Maks1-15 committed:
фз 10
Исправление внутренних тестов

Maks1-15 committed:
10 фз

Реализация CRUD для User and Film в InMemory

Maks1-15 committed:
Merge branch 'controllers-films-users' of https://github.com/Maks1-15/java-filmorate into controllers-films-users

# Conflicts:
#	pom.xml
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/model/User.java
#	src/main/java/ru/yandex/practicum/filmorate/repository/FilmRepository.java
#	src/main/java/ru/yandex/practicum/filmorate/repository/UserRepository.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java

Maks1-15 committed:
10 фз

Реализация CRUD для User and Film в InMemory
Добавлено:
- Логгирование
- Обработка запросов через ResponseEntity
- Обработка Exception

Maks1-15 committed:
10 фз

Реализация CRUD для User and Film в InMemory
Добавлено:
- Логгирование
- Обработка запросов через ResponseEntity
- Обработка Exception

Maks1-15 committed:
10 фз

Реализация CRUD для User and Film в InMemory
Добавлено:
- Логгирование
- Обработка запросов через ResponseEntity
- Обработка Exception

Maks1-15 committed:
10 фз

Реализация CRUD для User and Film в InMemory
Добавлено:
- Логгирование
- Обработка запросов через ResponseEntity
- Обработка Exception

Maks1-15 committed:
10 фз

Реализация CRUD для User and Film в InMemory
Добавлено:
- Логгирование
- Обработка запросов через ResponseEntity
- Обработка Exception

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,91 @@
+    private final FilmServiceImpl filmService;
+
+    @Autowired
+    public FilmController(FilmServiceImpl filmService) {
```
Чтобы максимально избавится от однотипного кода, очень удобно использовать [@RequiredArgsConstructor ](https://www.baeldung.com/spring-injection-lombok#:~:text=Constructor%20Injection%20With%20Lombok) из библиотеки `lombok`. Он автоматически заинжектит все нужные нам поля

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,29 @@
+
+import java.time.LocalDate;
+
+@Data
```
_Саморазвитие_
Лучше не использовать `@Data`, так как она несет за собой много скрытых проблем. Вот отличная статья [про применение lombok](https://medium.com/nuances-of-programming/lombock-%D1%85%D0%BE%D1%80%D0%BE%D1%88%D0%B5%D0%B5-%D0%B8-%D0%BF%D0%BB%D0%BE%D1%85%D0%BE%D0%B5-%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-4bd9cfdc5dd5)

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/validation/ReleaseDateValidation.java:
```diff
@@ -0,0 +1,20 @@
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = ReleaseDateValidator.class)
+public @interface ReleaseDateValidation {
```
Здорово, что реализовал кастомную аннотацию 👍

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/impl/FilmServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Optional;
+
+@Service
+public class FilmServiceImpl implements GenericRepository<Film> {
```
Сервис не должен наследоваться от репозитория

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/FilmRepositoryImpl.java:
```diff
@@ -0,0 +1,68 @@
+        films.remove(id);
+    }
+
+    private boolean existsByTitle(String title) {
```
Данного условия в тз нет, его нужно удалить

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,67 @@
+        users.remove(id);
+    }
+
+    private boolean existsByEmail(String email) {
```
Данного условия в тз нет, его нужно удалить

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -1,13 +0,0 @@
-import org.springframework.boot.test.context.SpringBootTest;
-
-@SpringBootTest
-class FilmorateApplicationTests {
```
Данный тест нужно вернуть, а также реализовать тестирование из тз:

```
Тестирование
Добавьте unit-тесты для валидации моделей. Убедитесь, что она работает на граничных условиях.
Подсказка: на что обратить внимание при тестировании

Проверьте, что валидация не пропускает пустые или неверно заполненные поля. Посмотрите, как контроллер реагирует на пустой запрос.
```

afomkina reviewed (changes requested):
Ты хорошо постарался - видно, что вложено много сил, и код получился действительно достойным. Но сейчас в работе много дополнительного кода, который выходит за рамки текущего задания, и это немного мешает оценке по основным критериям. Также в ТЗ отдельно указано, что нужно написать unit-тесты для валидации моделей - их обязательно стоит добавить. На этом этапе лучше сосредоточиться именно на базовом функционале и покрытии тестами.

Maks1-15 committed:
10 фз

Исправление ошибок

Maks1-15 committed:
10 фз

Исправление ошибок

afomkina reviewed (changes requested):
Все еще не хватает тестов валидации. Для тестирования спринговой валидации можно воспользоваться [валидатором](https://stackoverflow.com/questions/29069956/how-to-test-validation-annotations-of-a-class-using-junit)

Maks1-15 committed:
10 фз

Добавлены тесты для контроллеров

VadimZharkov commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Test
+    public void testCreateFilmWithInvalidData() throws Exception {
```
Сложным путем пошел, можно было проще. Но принимается.
